### PR TITLE
WIP: sync L1 and L2 separately

### PR DESCRIPTION
**Pull request by Mirko-von-Leipzig**
~~This is a WIP to check whether I am doing this message passing system correctly.~~

~~Currently I've just added a separate L1 sync function, which does not yet connect to anything. But I am hoping to get feedback on if this is the correct way to go about things.~~

This PR adds a standalone L1 sync process which uses channels to communicate update events.

The process further separates out the core logic from the Ethereum interactions using another channel. This facilitates testing the core logic without requiring Ethereum.

### CODE REVIEW

File: crates/pathfinder/src/state/ethereum.rs [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+    Updates(Vec<StateUpdateLog>),
+    /// An L1 reorg was detected. Contains the reorg tail.
+    /// i.e. all blocks __up-to-and-including__ this value are now invalid.
+    Reorg(StarknetBlockNumber),
+}
+
+/// Syncs L1 state update logs. Emits [Events](Event) which should be handled
+/// to update storage and respond to queries.
+async fn sync_ethereum_state<Tr: Transport>(
+    transport: Web3<Tr>,
```
Comment (Mirko-von-Leipzig) [inline crates/pathfinder/src/state/ethereum.rs:34]: Should this also be a separate channel? So that the web3 thingies should also be part of a separate operation?

I guess I'm unclear on which things should be externalised or not.

Comment (Mirko-von-Leipzig) [reply crates/pathfinder/src/state/ethereum.rs]: Made this a separate channel.

**Review by CHr15F0x — COMMENTED** [review]
From what I've understood this is a way of separating L1 sync logic from the actual Ethereum log fetcher logic, which uses a real `Transport` implementation that in consequence utilizes a real api (which sometimes fails and/or is slow to respond, and ofc does not produce ad-hoc corner-case data on demand).

I think this should also be achievable by mocking `Web3` and `Transport`. The question remains how easy or hard would it be to actually use for example `mockall` to do this.

On one hand the disadvantage of separating entities by splitting them into separate `tokio::task`s is that the number of concurrent tasks the developer has to be aware of (together with the entire message flow graph) is increased artificially to achieve easiness of injecting test data. There is also the issue of task cancellation which is a subject of its own.

The disadvantage of mocking is that one has to get used to the available mock apis, their pitfalls and limitations. Also depending on the api being mocked some apis are easy and other are hard to mock, which only usually turns out once one attempts to actually write the tests using those mocks.

File: crates/pathfinder/src/state/ethereum.rs [hunk 2]
```diff
@@ -0,0 +1,808 @@
...
+/// Syncs L1 state update logs. Emits [Events](Event) which should be handled
+/// to update storage and respond to queries.
+pub async fn sync_ethereum_state(
+    tx_event: mpsc::Sender<Event>,
+    transport: Web3<Http>,
+    chain: Chain,
+    head: Option<StateUpdateLog>,
+) -> anyhow::Result<()> {
+    // Spawn the Ethereum command handling function.
+    let (tx_eth, rx_eth) = mpsc::channel(5);
```
[the remark this answers is not in the document]
Comment (Mirko-von-Leipzig) [reply crates/pathfinder/src/state/ethereum.rs:52]: No proper reason, apart from `1` is likely too small. `2` may be perfect but I didn't want to worry about it just yet.

File: crates/pathfinder/src/state/ethereum.rs [hunk 3]
```diff
@@ -0,0 +1,808 @@
...
+    let _eth_future = tokio::spawn(eth_future);
+
+    // The core sync logic implementation.
+    sync_ethereum_state_impl(tx_eth, tx_event, head).await
+}
+
+/// Sends [Events](Event) on its channel.
+///
+/// Main purpose is to simplify the code in [sync_ethereum_state_impl] by abstracting
+/// out the error mapping and oneshot channel receives.
```
[the remark this answers is not in the document]
Comment (Mirko-von-Leipzig) [reply crates/pathfinder/src/state/ethereum.rs:63]: Either works 🤷

**Comment by Mirko-von-Leipzig** [comment]
> From what I've understood this is a way of separating L1 sync logic from the actual Ethereum log fetcher logic, which uses a real `Transport` implementation that in consequence utilizes a real api (which sometimes fails and/or is slow to respond, and ofc does not produce ad-hoc corner-case data on demand).
> 
> I think this should also be achievable by mocking `Web3` and `Transport`. The question remains how easy or hard would it be to actually use for example `mockall` to do this.
> 
> On one hand the disadvantage of separating entities by splitting them into separate `tokio::task`s is that the number of concurrent tasks the developer has to be aware of (together with the entire message flow graph) is increased artificially to achieve easiness of injecting test data. There is also the issue of task cancellation which is a subject of its own.
> 
> The disadvantage of mocking is that one has to get used to the available mock libraries, their pitfalls and limitations. Also depending on the api being mocked some apis are easy and other are hard to mock, which only usually turns out once one attempts to actually write the tests using those mocks.

I think I'll try this again with `mockall` and see how it plays out.

**Comment by CHr15F0x** [comment]
> how it plays out.

Yep, literally :+1: It's only after you start mocking this stuff then you can realize if it's a pain or not :shrug: . The benefit here would be not having to think about queue sizes, cancellation (if ever), boundary conditions enforcement in tests, and having the code look like typical sequential blocking code which is the advantage of async-await syntax. I'm really curious how it plays out :eyes:

File: crates/pathfinder/src/state/ethereum_mockall.rs [hunk 4]
```diff
@@ -0,0 +1,635 @@
...
+                    }
+
+                    // Reqeuest the previous Starknet block update.
+                    let update = match event_sender
+                        .get_update(reorg_tail.block_number.decrement())
+                        .await
+                    {
+                        Ok(update) => update,
+                        Err(_exit) => return Ok(()),
+                    };
```
Comment (Mirko-von-Leipzig) [inline crates/pathfinder/src/state/ethereum_mockall.rs:176]: I can imagine this "event" (which is really a request for a database read), should also come from a separate database API instead of being part of this event queue.

This would also neaten things up a lot I think.

File: crates/pathfinder/src/core.rs [hunk 5]
```diff
@@ -153,3 +153,21 @@ pub struct EthereumTransactionIndex(pub u64);
...
+impl StarknetBlockNumber {
+    pub const GENESIS: StarknetBlockNumber = StarknetBlockNumber(0);
+
+    pub fn increment(&self) -> Self {
+        Self(self.0 + 1)
+    }
+
+    pub fn decrement(&self) -> Self {
+        Self(self.0 - 1)
+    }
```
Comment (koivunej) [inline crates/pathfinder/src/core.rs:166]: Earlier was about to suggest we just have 

```
impl Add<usize> for StarknetBlockNumber {
    type Output = StarknetBlockNumber;
    
    ...
}
```

And I guess we need the decrement here as well. It would be a nice place to make sure these are checked_add and checked_sub.

File: crates/pathfinder/src/ethereum/log/fetch/forward.rs [hunk 6]
```diff
@@ -38,22 +38,30 @@ impl<T> LogFetcher<T>
...
         }
     }
 
+    pub fn set_head(&mut self, head: Option<T>) {
+        self.head = head;
+    }
+
+    pub fn head(&self) -> &Option<T> {
+        &self.head
+    }
```
Comment (koivunej) [inline crates/pathfinder/src/ethereum/log/fetch/forward.rs:63]: Dunno if clippy is saying anything on this, but in general idiomaticness `Option<&T>` as in `self.head.as_ref()` is more fun. It's as powerful as the `&Option<T>` for the caller, but you can just consume it.

File: crates/pathfinder/src/state/ethereum.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (koivunej) [reply crates/pathfinder/src/state/ethereum.rs:52]: I think it might be best to go with one when starting out. You can't really see these before-hand.

**Comment by Mirko-von-Leipzig** [comment]
I'm closing this, I'll likely re-open a completely different PR for this work which is a bit more organized.
